### PR TITLE
InteractionBox as EventedModel in Viewer

### PR DESCRIPTION
**Pull request by jojoelfe**
# Description
This is a refactor of #999 to create a viewer component that allows interactive transforms of layers or data within layers.

The InteractionBox component exposed as viewer.overlays.interaction_box and can be used by layers to make selections and transform items. Examples how it may be used by the points and image layer are included.

image_example

example_points

Compared to the original pull request the following changes were made:
- The model is now an EventedModel with most code placed into the vispy component
- InteractionBox is now a vispy overlay, similar to the scale bar or axes, and the model is part of the overlays class in the viewer, instead of the base layer
- The transforms now use the napari Affine class instead of skimage transforms

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# References

# How has this been tested?

- [x] Tested with included examples
- [ ] I'll write new unit tests

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Comment by jojoelfe** [comment]
@andy-sweet Here is the new refactor. It includes an example with the image layer, which will hopefully make things clearer. I haven't incorporated all of you reviews yet, but some got addressed with this refactor.

File: napari/_vispy/overlays/_interaction_box_constants.py
Comment (sofroniewn) [inline napari/_vispy/overlays/_interaction_box_constants.py]: This looks duplicated below do we need another copy?

File: napari/_vispy/overlays/interaction_box.py
Comment (sofroniewn) [inline napari/_vispy/overlays/interaction_box.py]: Do we need to add the mouse callbacks to the layer? Can we have them go on the viewer? Maybe can they be defined alongside the interaction box model? I don't love the vispy object getting so complicated. In general we've tried to define mouse functions along side the models and not inside the vispy objects

File: napari/components/interaction_box.py
Comment (sofroniewn) [inline napari/components/interaction_box.py:8]: This is very nice

File: napari/components/overlays.py
Comment (sofroniewn) [inline napari/components/overlays.py:7]: I like this too

File: napari/components/viewer_model.py
Comment (sofroniewn) [inline napari/components/viewer_model.py:119]: I like this, in the future we can refactor to include more overlays in here. Thoughts on that @jni? Shouldn't come in this PR

**Review by sofroniewn — COMMENTED** [review]
This is coming along really nicely. My main request would probably be to pull the mouse function stuff out of the vispy file if possible, and maybe put in its own file that lives next to the interaction box on the "model" side of the code base - maybe a private `_interaction_box_mouse_bindings.py` that would match the pattern here https://github.com/napari/napari/tree/main/napari/layers/points

File: napari/_vispy/overlays/_interaction_box_constants.py
Comment (jojoelfe) [reply napari/_vispy/overlays/_interaction_box_constants.py]: Yes, that's duplicated. It is not really needed for the model, so I removed it from there.

File: napari/_vispy/overlays/interaction_box.py
Comment (jojoelfe) [reply napari/_vispy/overlays/interaction_box.py]: Sorry, they already were on the viewer, I just did not rename the argument (fixed).

The problem with having the mouse events in the model is that the model than needs to be aware of the viewer object. My impression, mainly reading #1325, was that the models should remain very simple and unaware of implementation detail. I agree that there is now a lot of code in the vispy overlay. Maybe I could split it in a "Viewer" component that only handles rendering and a "Controller" component that handles the mouse events?

Comment (jojoelfe) [reply napari/_vispy/overlays/interaction_box.py]: Also, is there a way to remove callbacks from the callback list? There is no reason why the InteractionBox should be listening to events when it is not active. Right now the callback will just terminate immediately when not active.

Comment (sofroniewn) [reply napari/_vispy/overlays/interaction_box.py]: hmm yeah, we don't want this to have to know about the viewer. looking through code here it looks like the only viewer attribute that gets read is `viewer.dims.ndisplay`, so maybe there's a way to just pass that information and then you don't need to pass around the viewer as much? Maybe @andy-sweet can advise here too (though I know he's a little less familiar with this part of the code too). Otherwise I can look in more detail later maybe

Comment (jojoelfe) [reply napari/_vispy/overlays/interaction_box.py]: The viewer.dims.ndisplay is not really that essential, because for now this only works in 2D anyway. The main issue is appending the callbacks to the mouse-callback lists of the viewer object.

Comment (sofroniewn) [reply napari/_vispy/overlays/interaction_box.py]: > The main issue is appending the callbacks to the mouse-callback lists of the viewer object.

Ok yeah, we should be able to figure out a different way to do that, so the logic of adding/ removing the callback happens inside the viewer, but the callback itself is defined in a file like `_interaction_box_mouse_bindings.py`. Maybe easier said then done, but that would be ideal I think

Comment (jojoelfe) [reply napari/_vispy/overlays/interaction_box.py]: Ok, I moved the mouse binding code into a class called InteractionBoxMouseBindings that lives in `_interaction_box_mouse_bindings.py`. It gets instanced in `qt_viewer.py` together with the vispy class.

**Review by sofroniewn — APPROVED** [review]
Ok that's great - I love this! I'm approving this now, I think it is good to go, but maybe @andy-sweet can give it one more review before merge. Maybe @jni or @tlambert03 wants to take a quick look before merge too / give it a play around with.

I think it would be nice as a follow up PR if we could now start dropping some of the layer specific bounding box code (like in points and shapes) and reuse this overlay visual - that would show to me that we really added this in the right way. But that work can start after this merges.

Thanks also for your patience here, I know it took a long time to get the functionality merged!!

File: napari/components/interaction_box.py
Comment (andy-sweet) [inline napari/components/interaction_box.py]: Nit: no need for the `else` here.

**Review by tlambert03 — APPROVED** [review]
this is a beautiful PR!  So easy to follow, very well organized.  thanks!

the only minor suggestion/request I might have (and this can come in a followup) is that it'd be nice to get to the "common" use case of an interaction box appropriate for a given layer faster.  For example, the examples themselves (which should represent how we really want an end-user to _use_ these features) expose quite a few of the underlying attributes of the interaction box. 

```py
viewer.overlays.interaction_box.points = viewer.layers.selection.active.extent.world
viewer.overlays.interaction_box.show = True
viewer.overlays.interaction_box.show_vertices = True
viewer.overlays.interaction_box.show_handle = True
viewer.overlays.interaction_box.allow_new_selection = False
viewer.overlays.interaction_box.events.transform_drag.connect(on_transform_changed_drag)
```

for fun, I happened to run that example step by step in IPython, and could see that they really kind of needed to _all_ be run for it to make much sense.  I _do_ like very much how clean the interaction box model is, and that it has all of these toggles, but it would be nice to present a slightly more dumbed convenience for the common usage scenarios, like `viewer.add_interaction_box()` or something that gets me to a place where I'm interacting in less than 5 steps.

it's probably not _that_ simple to do that, but that was my first thought here. Anyway, love the model and the implementation!  LGTM!

**Comment by sofroniewn** [comment]
>  I do like very much how clean the interaction box model is, and that it has all of these toggles, but it would be nice to present a slightly more dumbed convenience for the common usage scenarios, like viewer.add_interaction_box() or something that gets me to a place where I'm interacting in less than 5 steps.

yeah, that's good feedback @tlambert03 - i agree. i think we can revisit after merge

File: napari/components/_interaction_box_mouse_bindings.py
Comment (andy-sweet) [inline napari/components/_interaction_box_mouse_bindings.py:7]: This is the same as `inside_boxes` in `_shapes_utils.py` right? It would be good to find a way to import/share this function before merging.

File: napari/components/_interaction_box_mouse_bindings.py
Comment (andy-sweet) [inline napari/components/_interaction_box_mouse_bindings.py]: Avoid leaving commented code. Either uncomment it, remove it, or if you really think this might be used in the future add a TODO with an implementation comment that describes/explains any current problems.

File: napari/components/interaction_box.py
Comment (andy-sweet) [inline napari/components/interaction_box.py:69]: This is very similar to the code a few lines above. Consider defining a helper function called `make_box(left, top, right, bottom)` or similar to avoid code duplication.

File: napari/components/_interaction_box_mouse_bindings.py
Comment (andy-sweet) [inline napari/components/_interaction_box_mouse_bindings.py]: Avoid commented code. See other comment for possible fixes.

File: napari/components/_interaction_box_mouse_bindings.py
Comment (andy-sweet) [inline napari/components/_interaction_box_mouse_bindings.py]: Avoid commented code. See other comment for possible fixes.

File: napari/components/_interaction_box_constants.py
Comment (andy-sweet) [inline napari/components/_interaction_box_constants.py:4]: Thanks for the explanatory comments - I think these were added from the last PR!

**Review by andy-sweet — COMMENTED** [review]
Looks great! I just had a few minor comments, small refactor options, and nits to pick.

Also wanted to check that by default the interaction box is not enabled, and that it's only by setting things up as in the example that it is enabled, right? I found a few issues/bugs on the example, but we can merge this and fix those later if it's not enabled by default.

File: napari/components/interaction_box.py
Comment (jojoelfe) [reply napari/components/interaction_box.py]: Fixed

File: napari/components/_interaction_box_mouse_bindings.py
Comment (jojoelfe) [reply napari/components/_interaction_box_mouse_bindings.py:7]: Yes, the plan was to remove the redundant code in a later PR where the shape layer will use the interaction box

File: napari/components/_interaction_box_constants.py
Comment (jojoelfe) [reply napari/components/_interaction_box_constants.py:4]: Sure, I think I addressed most of your comments from the last PR.

File: napari/components/interaction_box.py
Comment (jojoelfe) [reply napari/components/interaction_box.py:69]: I simplified it and put a comment there to explain the special case

**Comment by jojoelfe** [comment]
> Also wanted to check that by default the interaction box is not enabled, and that it's only by setting things up as in the example that it is enabled, right?

Yes, by default it is disabled and should not interfere in anything. 

> but it would be nice to present a slightly more dumbed convenience for the common usage scenarios, like `viewer.add_interaction_box()` or something that gets me to a place where I'm interacting in less than 5 steps.

I agree, especially for the use case where the box is used to control the transformation of a layer it should be possible to do everything in a simple helper function, which could either be in the interaction box or in the base layer class.

File: napari/components/interaction_box.py
Comment (andy-sweet) [reply napari/components/interaction_box.py:69]: Works for me!

**Review by andy-sweet — APPROVED** [review]

**Comment by sofroniewn** [comment]
Ok, I think this is done and ready for merge. I'll leave it to tomorrow in case @jni wants to give it a look, but as it is off by default I feel pretty good about merging it and then having a round of follow up PRs where we
- start using the interaction box for Points and Shapes (and get rid of the interaction boxes inside those layers)
- add some convenience APIs for creating an interaction box
- expose interaction box functionality via the GUI

Thanks all for the work here!!

**Review by alisterburt — APPROVED** [review]
Wow! Awesome PR (and so happy to see another cryo-ET-er on the napari repo! 🙂)

Mainly commenting to cc @brisvag and @kevinyamauchi as we were recently discussing overlays

**Comment by brisvag** [comment]
Nice! This plays right into our plans to unify overlays, and already does some of it. Great work @jojoelfe!

**Comment by sofroniewn** [comment]
Looks like this has gotten some more +1s, I'm going to merge now!!

**Comment by VolkerH** [comment]
This looks coll, saw that this is already merged, but just wondering whether it is possible to set aspect ratio constraints on the box @sofroniewn ?

**Comment by jojoelfe** [comment]
In theory yes, but in practice the keyboard events are not connected yet. A workaround at the moment is 

`viewer.window.qt_viewer.interaction_box_mousebindings._fixed_aspect = 1.0`

After that dragging on the corners maintains the aspect ratio.

**Comment by sofroniewn** [comment]
> In theory yes, but in practice the keyboard events are not connected yet.

I'd support a followup PR for that

**Pull request merged**
